// @tlmz/server/src/room-manager.ts
import { appendFileSync, mkdirSync } from 'node:fs';
import { join } from 'node:path';
import { applyAction, availableActions, makeGame, publicStateFor, type ClientAction, type GameLog, type GameState } from '@tlmz/rules';
import type { WebSocket } from 'ws';

type Client = { ws: WebSocket; roomId?: string; seat?: number; seats?: Set<number>; name?: string };
type SavedLog = GameLog & { at: string; roomId: string; round: number };
type Room = { id: string; game?: GameState; clients: Map<number, Client>; nextSeed: number; history: GameState[]; logs: SavedLog[]; logFile: string };
const logDir = join(process.cwd(), 'logs');
mkdirSync(logDir, { recursive: true });
function roomLogFile(roomId: string) { return join(logDir, `room-${roomId}.jsonl`); }
function cloneGame(game: GameState): GameState { return JSON.parse(JSON.stringify(game)); }
function nextDealerSeat(prev: GameState) {
  if (prev.settlement?.selfDraw) return prev.settlement.winnerSeat;
  if (prev.settlement?.discarderSeat != null) return prev.settlement.discarderSeat;
  return prev.winnerSeat ?? ((prev.dealerSeat + 1) % 4);
}
function saveLogs(room: Room, game: GameState, entries: GameLog[]) {
  if (!entries.length) return;
  const saved = entries.map(entry => ({ ...entry, at: new Date().toISOString(), roomId: room.id, round: game.round }));
  room.logs.push(...saved);
  appendFileSync(room.logFile, saved.map(e => JSON.stringify(e)).join('\n') + '\n', 'utf8');
}
function autoDrawIfUncontested(game: GameState) {
  if (game.phase !== 'playing') return;
  if (game.turnDrawn || game.pendingXiSupplement) return;
  const actions = availableActions(game, game.currentSeat);
  if (actions.includes('DRAW')) applyAction(game, { type: 'DRAW', seat: game.currentSeat });
  else if (actions.includes('FENZHANG')) applyAction(game, { type: 'FENZHANG', seat: game.currentSeat });
}

export class RoomManager {
  rooms = new Map<string, Room>();
  createRoom(): string {
    const id = Math.floor(100000 + Math.random() * 900000).toString();
    this.rooms.set(id, { id, clients: new Map(), nextSeed: Date.now() % 1000000, history: [], logs: [], logFile: roomLogFile(id) });
    return id;
  }
  joinRoom(client: Client, roomId: string, name: string) {
    let room = this.rooms.get(roomId);
    if (!room) { room = { id: roomId, clients: new Map(), nextSeed: Date.now() % 1000000, history: [], logs: [], logFile: roomLogFile(roomId) }; this.rooms.set(roomId, room); }
    const occupied = new Set(room.clients.keys());
    let seat = 0; while (occupied.has(seat) && seat < 4) seat++;
    if (seat >= 4) throw new Error('房间已满');
    client.roomId = roomId; client.seat = seat; client.name = name; client.seats ??= new Set<number>(); client.seats.add(seat); room.clients.set(seat, client);
    if (!room.game) { room.game = makeGame({ seed: room.nextSeed, dealerSeat: 0, roomId }); room.game.phase = 'waiting'; saveLogs(room, room.game, room.game.log); }
    room.game.players[seat].name = name; room.game.players[seat].connected = true;
    this.broadcast(roomId);
    return { roomId, seat, controlledSeats: [...client.seats] };
  }
  start(roomId: string) {
    const room = this.get(roomId); if (room.clients.size < 4) throw new Error('四个人到齐后才能开始游戏'); room.history = [];
    const names = [0,1,2,3].map(seat => room.clients.get(seat)?.name ?? `玩家${seat+1}`);
    const connected = [0,1,2,3].map(seat => room.clients.has(seat));
    const scores = room.game?.scores ?? [1000,1000,1000,1000];
    room.game = makeGame({ seed: ++room.nextSeed, dealerSeat: room.game?.dealerSeat ?? 0, roomId, scores, round: room.game?.round ?? 1, phase: 'betting', names, connected });
    saveLogs(room, room.game, room.game.log);
    this.broadcast(roomId);
  }
  action(roomId: string, client: Client, action: ClientAction) {
    const room = this.get(roomId); if (!room.game) throw new Error('游戏未开始');
    const controlledSeats = client.seats ?? (client.seat != null ? new Set([client.seat]) : new Set<number>());
    if (!controlledSeats.has(action.seat)) throw new Error('不能操作未控制的座位');
    room.history.push(cloneGame(room.game)); if (room.history.length > 50) room.history.shift();
    const beforeLogLength = room.game.log.length;
    try { applyAction(room.game, action); autoDrawIfUncontested(room.game); } catch (e) { room.history.pop(); throw e; }
    saveLogs(room, room.game, room.game.log.slice(beforeLogLength));
    if (room.game.phase === 'settlement' && room.game.readyNext.every(Boolean)) {
      const prev = room.game;
      const names = prev.players.map(p => p.name);
      const connected = prev.players.map(p => !!p.connected);
      const dealerSeat = nextDealerSeat(prev);
      room.history = [];
      room.game = makeGame({ seed: ++room.nextSeed, dealerSeat, roomId, scores: prev.scores, round: prev.round + 1, phase: 'betting', names, connected });
      saveLogs(room, room.game, room.game.log);
    }
    this.broadcast(roomId);
  }
  undo(roomId: string) {
    const room = this.get(roomId); const prev = room.history.pop(); if (!prev) throw new Error('没有可撤回的操作'); room.game = prev; this.broadcast(roomId);
  }
  disconnect(client: Client) {
    if (client.roomId != null) {
      const room = this.rooms.get(client.roomId);
      for (const seat of client.seats ?? (client.seat != null ? [client.seat] : [])) { room?.clients.delete(seat); if (room?.game) room.game.players[seat].connected = false; }
      if (client.roomId) this.broadcast(client.roomId);
    }
  }
  getLogs(roomId: string) { const room = this.get(roomId); return room.logs; }
  private get(roomId: string): Room { const room = this.rooms.get(roomId); if (!room) throw new Error('房间不存在'); return room; }
  broadcast(roomId: string) {
    const room = this.rooms.get(roomId); if (!room?.game) return;
    const game = room.game;
    const sent = new Set<Client>();
    for (const [, client] of room.clients) {
      if (sent.has(client)) continue; sent.add(client);
      const controlledSeats = [...(client.seats ?? (client.seat != null ? [client.seat] : []))];
      const seat = client.seat ?? controlledSeats[0] ?? 0;
      const actionsBySeat = Object.fromEntries([0,1,2,3].map(s => [s, availableActions(game, s)]));
      const payload = { type: 'STATE', seat, controlledSeats, state: publicStateFor(game, seat), actions: availableActions(game, seat), actionsBySeat, connectedCount: room.clients.size, canStart: room.clients.size === 4, canUndo: room.history.length > 0 };
      client.ws.send(JSON.stringify(payload));
    }
  }
}
export type { Client };
