// @tlmz/rules/src/state.ts
import { buildWall, dealInitialHands, shuffleWall } from './wall.js';
import { getFishTile } from './fish.js';
import { sortTiles, type TileCode } from './tiles.js';

export type MeldType = 'chi' | 'peng' | 'mingGang' | 'anGang' | 'xi' | 'dingZhang';
export type Meld = { type: MeldType; tiles: TileCode[]; fromSeat?: number; name?: string };
export type PlayerState = { seat: number; name: string; hand: TileCode[]; melds: Meld[]; discarded: TileCode[]; connected?: boolean };
export type GamePhase = 'waiting' | 'betting' | 'playing' | 'responding' | 'fenzhang' | 'settlement' | 'finished';
export type GameLog = { type: string; [key:string]: unknown };
export type RoundBet = { zha: boolean; buyFish: number; ready: boolean };
export type ScoreDelta = { seat: number; before: number; delta: number; after: number; detail: string };
export type Settlement = { winnerSeat: number; selfDraw: boolean; baseScore: number; fishTotal: number; ownFish: number; boughtFishTotal: number; zhaSeats: number[]; discarderSeat?: number; deltas: ScoreDelta[] };
export type ResponseWindow = { discardSeat: number; tile: TileCode; source: 'normal' | 'xi'; passed: number[] };

export type GameState = {
  roomId: string;
  phase: GamePhase;
  dealerSeat: number;
  currentSeat: number;
  round: number;
  scores: number[];
  baseScore: number;
  bets: RoundBet[];
  readyNext: boolean[];
  settlement?: Settlement;
  turnDrawn: boolean;
  xiWindowOpen: boolean;
  openingXiDone: boolean;
  pendingXiSupplement: boolean;
  wall: TileCode[];
  revealedFishTile: TileCode;
  fishTile: TileCode;
  players: PlayerState[];
  log: GameLog[];
  lastDiscard?: { seat:number; tile:TileCode; source:'normal'|'xi' };
  response?: ResponseWindow;
  drawnTile?: {seat:number; tile:TileCode};
  winnerSeat?: number;
};

export function makeGame(opts: { seed?: number; dealerSeat?: number; roomId?: string; scores?: number[]; round?: number; phase?: GamePhase; names?: string[]; connected?: boolean[] } = {}): GameState {
  const dealerSeat = opts.dealerSeat ?? 0;
  const shuffled = shuffleWall(buildWall(), opts.seed ?? Date.now());
  const revealedFishTile = shuffled[0];
  const fishTile = getFishTile(revealedFishTile);
  const dealt = dealInitialHands(shuffled.slice(1), dealerSeat);
  const players = dealt.hands.map((h, seat) => ({
    seat,
    name: opts.names?.[seat] ?? `玩家${seat + 1}`,
    hand: sortTiles(h),
    melds: [],
    discarded: [],
    connected: opts.connected?.[seat] ?? false,
  }));

  return {
    roomId: opts.roomId ?? 'local',
    phase: opts.phase ?? 'playing',
    dealerSeat,
    currentSeat: dealerSeat,
    round: opts.round ?? 1,
    scores: opts.scores ? [...opts.scores] : [1000, 1000, 1000, 1000],
    baseScore: 1,
    bets: Array.from({ length: 4 }, () => ({ zha: false, buyFish: 0, ready: false })),
    readyNext: [false, false, false, false],
    turnDrawn: true,
    xiWindowOpen: true,
    openingXiDone: false,
    pendingXiSupplement: false,
    wall: dealt.wall,
    revealedFishTile,
    fishTile,
    players,
    log: [{ type: 'start', dealerSeat, revealedFishTile, fishTile }],
  };
}

export function publicStateFor(state: GameState, seat?: number) {
  const showAllHands = true; // 测试版本：默认明牌，方便四个座位一起调试整局流程。
  return { ...state, players: state.players.map(p => ({ ...p, hand: showAllHands || seat === p.seat ? p.hand : p.hand.map(() => 'BACK' as const) })) };
}
